File: dailys/advent_four.py
import re

def check_x(mat, i, j):
    # Construct the string from the positions around 'A'
    x_string = (
        mat[i-1][j-1] +  # Top Left
        mat[i][j] +      # Center A
        mat[i+1][j+1] +  # Bottom Right
        mat[i-1][j+1] +  # Top Right
        mat[i][j] +      # Center A again (for symmetry)
        mat[i+1][j-1]    # Bottom Left
    )
    
    # Check if the constructed string matches any valid X-MAS patterns
    return x_string in ["MASMAS", "SAMSAM", "SAMMAS", "MASSAM"]

def count_x_mas_patterns(mat):
    x_mas_count = 0
    rows, cols = len(mat), len(mat[0])
    
    for i in range(1, rows - 1):  # Avoid outermost rows
        for j in range(1, cols - 1):  # Avoid outermost columns
            if mat[i][j] == "A":
                x_mas_count += int(check_x(mat, i, j))
    
    return x_mas_count

def part_2(input_lines):
    mul_search = re.compile(r"mul\(([0-9]{1,3}),([0-9]{1,3})\)")
    
    merged_data = "".join([line.strip() for line in input_lines])
    
    dirty_processing = [segment.split("don't()")[0] for segment in merged_data.split("do()")]
    
    results = mul_search.findall("".join(dirty_processing))
    
    sum_of_multiples = sum(int(x) * int(y) for x, y in results)

    # Extracting the grid from input_lines (assuming each line is a row)
    matrix = [list(line.strip()) for line in input_lines if line.strip()] 

    # Count X-MAS patterns in the extracted grid
    x_mas_count = count_x_mas_patterns(matrix)

    return sum_of_multiples, x_mas_count

# Example usage:
with open("sample.txt", "r") as file:
    input_lines = file.readlines()

sum_of_multiples, x_mas_count = part_2(input_lines)
print(f"Sum of multiples: {sum_of_multiples}, X-MAS patterns found: {x_mas_count}")

File: advent_php/advent_four.php
<?php

function check_x($mat, $i, $j) {
    // Construct the string from the positions around 'A'
    $x_string = 
        $mat[$i-1][$j-1] .  // Top Left
        $mat[$i][$j] .      // Center A
        $mat[$i+1][$j+1] .  // Bottom Right
        $mat[$i-1][$j+1] .  // Top Right
        $mat[$i][$j] .      // Center A again (for symmetry)
        $mat[$i+1][$j-1];   // Bottom Left
    
   
    return in_array($x_string, ["MASMAS", "SAMSAM", "SAMMAS", "MASSAM"]);
}

function count_x_mas_patterns($mat) {
    $x_mas_count = 0;
    $rows = count($mat);
    $cols = count($mat[0]);
    
    for ($i = 1; $i < $rows - 1; $i++) {  
        for ($j = 1; $j < $cols - 1; $j++) { 
            if ($mat[$i][$j] == "A") {
                $x_mas_count += check_x($mat, $i, $j) ? 1 : 0;
            }
        }
    }
    
    return $x_mas_count;
}

function part_2($input_lines) {
    $mul_search = '/mul\(([0-9]{1,3}),([0-9]{1,3})\)/';
    
    $merged_data = implode("", array_map('trim', $input_lines));
    
    $dirty_processing = array_map(function($segment) {
        return explode("don't()", $segment)[0];
    }, explode("do()", $merged_data));
    
    preg_match_all($mul_search, implode("", $dirty_processing), $results, PREG_SET_ORDER);
    
    $sum_of_multiples = array_reduce($results, function($carry, $item) {
        return $carry + (int)$item[1] * (int)$item[2];
    }, 0);

   
    $matrix = array_map('str_split', array_filter(array_map('trim', $input_lines)));

    $x_mas_count = count_x_mas_patterns($matrix);

    return [$sum_of_multiples, $x_mas_count];
}

$input_lines = file('input.txt', FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
list($sum_of_multiples, $x_mas_count) = part_2($input_lines);

echo "Sum of multiples: " . $sum_of_multiples . "\n";
echo "X-MAS patterns count: " . $x_mas_count . "\n";

?>